uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/providers/IdentityProviderEndpointsAliasMockMvcTests.java

```diff
@@ -327,229 +373,351 @@ protected UpdateBase(final boolean aliasFeatureEnabled) {
+                private void shouldReject_AliasIdNotSetInPayload(
+                        final IdentityZone zone1,
+                        final IdentityZone zone2
+                ) throws Exception {
+                    final IdentityProvider<?> existingIdp = createIdpWithAlias(zone1, zone2);
+
+                    existingIdp.setAliasId(null);
+                    existingIdp.setName("some-new-name");
+                    shouldRejectUpdate(zone1, existingIdp, HttpStatus.UNPROCESSABLE_ENTITY);
+                }
+
+                @Test
+                void shouldAccept_ShouldFixDanglingRefByCreatingNewAlias_UaaToCustomZone() throws Exception {
+                    shouldAccept_ShouldFixDanglingRefByCreatingNewAlias(IdentityZone.getUaa(), customZone);
+                }
+
+                @Test
+                void shouldAccept_ShouldFixDanglingRefByCreatingNewAlias_CustomToUaaZone() throws Exception {
+                    shouldAccept_ShouldFixDanglingRefByCreatingNewAlias(customZone, IdentityZone.getUaa());
+                }
 
-                // try to create an alias for the IdP -> should fail because of the IdP's type
-                createdProvider.setAliasZid(zone2.getId());
-                shouldRejectUpdate(zone1, createdProvider, HttpStatus.UNPROCESSABLE_ENTITY);
+                private void shouldAccept_ShouldFixDanglingRefByCreatingNewAlias(final IdentityZone zone1, final IdentityZone zone2) throws Exception {
+                    final IdentityProvider<?> idp = createIdpWithAlias(zone1, zone2);
+
+                    // delete the alias IdP directly in the DB -> after that, there is a dangling reference
+                    deleteIdpViaDb(idp.getOriginKey(), zone2.getId());
+
+                    // update some other property on the original IdP
+                    idp.setName("some-new-name");
+                    final IdentityProvider<?> updatedIdp = updateIdp(zone1, idp);
+                    assertThat(updatedIdp.getAliasId()).isNotBlank().isNotEqualTo(idp.getAliasId());
+                    assertThat(updatedIdp.getAliasZid()).isNotBlank().isEqualTo(idp.getAliasZid());
+
+                    // check if the new alias IdP is present and has the correct properties
+                    final String id = updatedIdp.getAliasId();
+                    final Optional<IdentityProvider<?>> aliasIdp = readIdpFromZoneIfExists(zone2.getId(), id);
+                    assertThat(aliasIdp).isPresent();
+                    assertIdpReferencesOtherIdp(updatedIdp, aliasIdp.get());
+                    assertOtherPropertiesAreEqual(updatedIdp, aliasIdp.get());
+
+                    // check if both have the same non-empty relying party secret
+                    assertIdpAndAliasHaveSameRelyingPartySecretInDb(updatedIdp);
+
+                    // check if the returned IdP has a redacted relying party secret
+                    assertRelyingPartySecretIsRedacted(updatedIdp);
+                }
```

Comment:
Looks good. Leaving comments to track my progress as this is challenging to read through.